src/libsync/propagateuploadng.cpp

```diff
@@ -30,23 +30,93 @@
 #include <QNetworkAccessManager>
 #include <QFileInfo>
 #include <QDir>
+
 #include <cmath>
 #include <cstring>
 
 namespace OCC {
 
-QUrl PropagateUploadFileNG::chunkUrl(int chunk)
+QUrl PropagateUploadFileNG::chunkUrl(qint64 chunk)
 {
     QString path = QLatin1String("remote.php/dav/uploads/")
         + propagator()->account()->davUser()
         + QLatin1Char('/') + QString::number(_transferId);
     if (chunk >= 0) {
-        // We need to do add leading 0 because the server orders the chunk alphabetically
-        path += QLatin1Char('/') + QString::number(chunk).rightJustified(8, '0');
+        path += QLatin1Char('/') + QString::number(chunk);
     }
     return Utility::concatUrlPath(propagator()->account()->url(), path);
 }
 
+void PropagateUploadFileNG::slotZsyncSeedFinished(void* zs)
+{
+    zsync_unique_ptr<struct zsync_state> _zs((struct zsync_state*)zs, [](struct zsync_state* zs) {
+        zsync_end(zs);
+    });
+    {   /* And print how far we've progressed towards the target file */
+        long long done, total;
+
+        zsync_progress(_zs.get(), &done, &total);
+        qCInfo(lcZsyncPut).nospace() << "Done reading: "
+            << _item->_file << " " << fixed << qSetRealNumberPrecision(1) <<  (100.0f * done) / total
+            << "% of target seeded.";
+
+    }
+
+    QTemporaryFile tempF;
+    tempF.open();
+    std::string temp = tempF.fileName().toStdString();
+    const char * tmpfile = temp.c_str();
+    if (zsync_rename_file(_zs.get(), tmpfile) != 0) {
+        abortWithError(SyncFileItem::NormalError,  tr("Unable to rename temporary file."));
+        return;
+    }
+
+    /* Get a set of byte ranges that we need to complete the target */
+    int _nrange = 0;
+    off_t *_zbyterange = zsync_needed_byte_ranges(_zs.get(), &_nrange, 0);
+    if (!_zbyterange) {
+        abortWithError(SyncFileItem::NormalError, tr("Failed to get zsync byte ranges."));
+        return;
+    }
+
+    /* If we have no ranges then we have equal files and we are done */
+    if (_nrange == 0) {
+        free(_zbyterange);
+        finalize();
+        return;
+    }
+
+    int newSize = 0;
+    for (int i=0; i<_nrange; i++) {
+        ServerRangeInfo rangeinfo = { (quint64)_zbyterange[(2*i)], (quint64)_zbyterange[(2*i)+1] - (quint64)_zbyterange[(2*i)] + 1 };
+        if (rangeinfo.start < _item->_size) {
+            if (rangeinfo.start + rangeinfo.size > _item->_size)
+                rangeinfo.size = _item->_size - rangeinfo.start;
+            _serverRanges.append(rangeinfo);
+            newSize += rangeinfo.size;
+        }
+    }
+    if (_item->_size > (quint64)zsync_file_length(_zs.get())) {
+        ServerRangeInfo rangeinfo = { (quint64)zsync_file_length(_zs.get()), (quint64)_item->_size - (quint64)zsync_file_length(_zs.get()) };
+        _serverRanges.append(rangeinfo);
+        newSize += rangeinfo.size;
+    }
+
+    propagator()->reportFileTotal(*_item, newSize);
+
+    _fileSize = newSize;
```

Comment:
Maybe not call the member `_fileSize` but `_bytesToDownload` or something?

Reply:
sure, but doesn't `_bytesToUpload` make more sense here?